### DISCUSSION TITLE
Next.js keeps injecting the default favicon

### DISCUSSION THREAD

**Discussion by akoskm** [vercel/next.js#46415] [Help]
Hi!

I configured the metadata constant in `layout.tsx` as follows:

```
export const metadata = {
  title: 'Some title,
  description: 'Some description',
  icons: {
    icon: '/favicons/favicon.ico',
    shortcut: '/favicons/favicon-16x16.png',
    apple: '/favicons/favicon-32x32.png',
    other: {
      rel: 'apple-touch-icon-precomposed',
      url: '/favicons/apple-touch-icon.png',
    },
  },
}
```
I also tried `icon: { url: '/favicons/favicon.ico', sizes: 'any', type: 'image/x-icon' },` instead of having simply `icon: '/favicons/favicon.ico'`

I have all the icons in the `/favicons` folder.

The problem is that the default Vercel icons keep getting injected into `<meta>, overriding my custom favicon:

![image](https://user-images.githubusercontent.com/3111627/221358468-744562cc-dd0d-49a6-84bd-fcb5037b6c3c.png)

The weirdest thing is if I now go and replace `icon: '/favicons/favicon.ico'` with `icon: { url: '/favicons/favicon.ico', sizes: 'any', type: 'image/x-icon' },` Next.js does an auto-reload, and the custom favicon is visible until I don't reload the page while the default icon is also still injected. Once I reload the page, the default icon takes priority again.

Tried clearing the cache, private window, Brave, and Safari. It's the same behavior.

Side question: What are the best practices for using the `favicon`, `site.webmanifest`, etc files? I assume the steps described here https://github.com/vercel/next.js/discussions/21496 are outdated.

**Comment by chuckcg** [comment]
Hi @akoskm , I have the same issue. Did you find a workaround ?

**Reply by akoskm** [reply]
No, I haven't. I'll report back here once I figure this out!

**Reply by icyJoseph** [reply]
Think it was introduced about 2 weeks ago, that if you have favicon.ico under app, it uses that one:

- https://github.com/vercel/next.js/pull/46997

**Reply by chuckcg** [reply]
Indeed, as @icyJoseph pointed out, removing favicon.ico from the app directory fixed it for me @akoskm

**Comment by icyJoseph** [comment]
I think this happens when the `src/app/favicon.ico` file is present, not sure why the app dir bundler picks it up from there, but looks like that's the case.

**Comment by Megabyte-webdev** [comment]
> Hi!
> 
> I configured the metadata constant in `layout.tsx` as follows:
> 
> ```
> export const metadata = {
>   title: 'Some title,
>   description: 'Some description',
>   icons: {
>     icon: '/favicons/favicon.ico',
>     shortcut: '/favicons/favicon-16x16.png',
>     apple: '/favicons/favicon-32x32.png',
>     other: {
>       rel: 'apple-touch-icon-precomposed',
>       url: '/favicons/apple-touch-icon.png',
>     },
>   },
> }
> ```
> 
> I also tried `icon: { url: '/favicons/favicon.ico', sizes: 'any', type: 'image/x-icon' },` instead of having simply `icon: '/favicons/favicon.ico'`
> 
> I have all the icons in the `/favicons` folder.
> 
> The problem is that the default Vercel icons keep getting injected into `, overriding my custom favicon:
> 
> ![image](https://user-images.githubusercontent.com/3111627/221358468-744562cc-dd0d-49a6-84bd-fcb5037b6c3c.png)
> 
> The weirdest thing is if I now go and replace `icon: '/favicons/favicon.ico'` with `icon: { url: '/favicons/favicon.ico', sizes: 'any', type: 'image/x-icon' },` Next.js does an auto-reload, and the custom favicon is visible until I don't reload the page while the default icon is also still injected. Once I reload the page, the default icon takes priority again.
> 
> Tried clearing the cache, private window, Brave, and Safari. It's the same behavior.
> 
> Side question: What are the best practices for using the `favicon`, `site.webmanifest`, etc files? I assume the steps described here #21496 are outdated.

There is always a next favicon.ico in the app directory of your project, just replace your favicon with that.

**Comment by tadanobutubutu** [comment]
### Possible Cause and Solution

This common issue in the Next.js App Router often stems from how static metadata files are prioritized over the `metadata` object in your layout.

#### 1. Check for `app/favicon.ico`
Next.js automatically detects and uses a `favicon.ico` (or `.png`/`.svg`) if it's placed directly in the `app/` directory. Even if you define icons in your `layout.tsx` metadata constant, the physical file in the `app/` folder often takes precedence or causes a conflict where both are injected, and browsers might default to the root one.

**Solution:** Delete `app/favicon.ico` if you want to use the one defined in your `metadata` object, or replace it with your custom icon.

#### 2. Browser Caching
Favicons are aggressively cached. Even after a server-side fix, your browser might still show the old one.
**Solution:** Try adding a version query string to your icon path to force a refresh:
```typescript
icon: '/favicons/favicon.ico?v=1',
```

#### 3. Absolute vs Relative Paths
Ensure your `/favicons` directory is inside the `public/` directory (i.e., `public/favicons/`). In the metadata object, the path should be relative to the public root.

#### 4. Explicit `rel` attribute
Sometimes explicitly defining the icon as an object helps:
```typescript
icons: {
  icon: [{ url: '/favicons/favicon.ico', rel: 'icon' }],
  apple: [{ url: '/favicons/apple-touch-icon.png', rel: 'apple-touch-icon' }],
},
```

Hope this helps resolve the injection conflict!
